core/src/main/java/org/frankframework/pipes/IfPipe.java

```diff
@@ -0,0 +1,277 @@
+
+		final MediaType mediaType;
+
+		DefaultMediaType(MediaType mediaType) {
+			this.mediaType = mediaType;
+		}
+	}
+
+	private TransformerPool transformerPool;
+	private String jsonPathExpression = null;
+	private String regex = null;
+	private String xpathExpression = null;
+	private int xsltVersion = XmlUtils.DEFAULT_XSLT_VERSION;
+
+	@Override
+	public void configure() throws ConfigurationException {
+		super.configure();
+
+		if (StringUtils.isNotEmpty(xpathExpression)) {
+			try {
+				transformerPool = TransformerPool.getInstance(makeStylesheet(xpathExpression, expressionValue), xsltVersion, this);
+			} catch (TransformerConfigurationException e) {
+				throw new ConfigurationException("could not create transformer from xpathExpression [" + xpathExpression + "], target expressionValue [" + expressionValue + "]", e);
+			}
+		}
+	}
+
+	@Override
+	public PipeRunResult doPipe(Message message, PipeLineSession session) throws PipeRunException {
+		MimeType mimeType = getMimeType(message);
+
+		// Determine the media type and if the correct *pathExpression was provided
+		if (jsonPathExpression != null && !MediaType.APPLICATION_JSON.equals(mimeType)
+				|| xpathExpression != null && !MediaType.APPLICATION_XML.equals(mimeType)) {
+			throw new PipeRunException(this, "Incorrect pathExpression provided for given mediaType + " + mimeType);
+		}
+
+		String forward = determineForward(message, session);
+
+		log.debug("determined forward [{}]", forward);
+
+		PipeForward pipeForward = findForward(forward);
+		if (pipeForward == null) {
+			throw new PipeRunException(this, "cannot find forward or pipe named [" + forward + "]");
+		}
+
+		log.debug("resolved forward [{}] to path [{}]", forward, pipeForward.getPath());
+		return new PipeRunResult(pipeForward, message);
+	}
+
+	private MimeType getMimeType(Message message) {
+		MimeType computedType = MessageUtils.computeMimeType(message);
+
+		if (computedType != null) {
+			// check if computedType is one of JSON or XML, if not, fall through
+			boolean matchesOneOfSupportedTypes = Arrays.stream(DefaultMediaType.values())
+					.anyMatch(supportedType -> supportedType.mediaType.equals(computedType));
+
+			if (matchesOneOfSupportedTypes) {
+				return computedType;
+			}
+		}
+
+		// Default to
+		return defaultMediaType.mediaType;
+	}
+
+	/**
+	 * Based on the pipe settings, tries to evaluate the expression and determine the correct forward
+	 */
+	private String determineForward(Message message, PipeLineSession session) throws PipeRunException {
+		String inputString = getInputString(message);
+
+		if (xpathExpression != null && transformerPool != null) {
+			try {
+				Map<String, Object> parameterValues = null;
+				ParameterList parameterList = getParameterList();
+				if (!parameterList.isEmpty()) {
+					parameterValues = parameterList.getValues(message, session, namespaceAware).getValueMap();
+				}
+				return transformerPool.transform(inputString, parameterValues, namespaceAware);
```

Comment:
ik snap hem wel, maar let wel op dat je nu op 2 plekken de 'if- check' doet en hier nu niet de `expressionValue` gebruikt. 
Mogelijk is het beter het resultaat terug te geven en de expressionValue (generiek) elders doet.

Reply:
Ik heb vooral niet aan de xml gebaseerde code gezeten, ook dit komt zo uit de xmlif. Ik zie ook niet helemaal wat je met de 'if check' bedoelt. Kan je dat zo even toelichten?